### PR TITLE
Clarify API for GP approximations

### LINKED ISSUE #221: Add API for (log) marginal likelihood approximation
AbstractGPs currently defines `elbo`, but other approximations (e.g. Laplace, EP) also provide approximations to the (log) marginal likelihood that can be used for hyperparameter optimisation. Would be great to have a unified API function for that (for VFE, SVGP, etc it could simply forward to `elbo`). What do you think? Finding a decent name seems the most complicated aspect of this:
- `approx_lml` (short but a bit cryptic)?
- `approx_log_marginal_likelihood` (explicit but so long)?
- `approx_log_marglik` (the compromise that makes everyone unhappy)?
- `approx_log_evidence`?

### PR DESCRIPTION
**Pull request by st--**
**Summary**
AbstractGPs.jl should define the API we expect across the ecosystem. Currently, this is not the case for GP approximations, even though they're already introduced in this package (`VFE` for sparse GPR).

**Proposed changes**
Expose `posterior` and `approx_log_evidence` as part of the API with nice generic docstrings explaining their purpose.

**Related issues**
Resolves #221 

Also see #223, #241, #318, maybe #319

**What alternatives have you considered?**
Status quo: some definitions in ApproximateGPs.jl, but people not realizing that you could just use a three-parameter form for `posterior`, leading to more reimplementation of base FiniteGPs than would be necessary (e.g. https://github.com/SebastianCallh/IterGP.jl/blob/AbstractGPs-interface/src/gp.jl)

### CODE REVIEW

**Comment by st--** [comment]
> Should there be an `AbstractGPApproximation` type or so? That other packages can/ should use?

Hm... where would that be useful? Is there anything that you would be able to do with _any_ approximation, without knowing what approximation it is (i.e. it'd be necessary as a common base type in the hierarchy)? Or just as a "type hint" (i.e. it might be better to just improve the docs)?

**Comment by st--** [comment]
Open question from https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/issues/221#issuecomment-1521960655:

should we add a `struct ExactGP end` "approximation" type to support a consistent API regardless of whether one is using exact GPR or a sparse approximation such as VFE/DTC?

**Comment by Crown421** [comment]
> Open question from [#221 (comment)](https://github.com/JuliaGaussianProcesses/AbstractGPs.jl/issues/221#issuecomment-1521960655):
> 
> should we add a `struct ExactGP end` "approximation" type to support a consistent API regardless of whether one is using exact GPR or a sparse approximation such as VFE/DTC?

I think this makes sense to me, with appropriate aliases for the current two argument form. 

Re the above, I was thinking about a type hint, need to think about whether such a type might also be useful elsewhere.

**Comment by Crown421** [comment]
One way I can think of where a `AbstractApproximation` type might be useful in starting a more general hierarchy. 
Next level down would be the methods mentioned here: https://github.com/JuliaGaussianProcesses/ApproximateGPs.jl/issues/84, might might be some supertype of `VFE`, `DTC` (possibly others in the future like FITC)

File: src/exact_gpr_posterior.jl [hunk 1]
```diff
@@ -3,6 +3,12 @@ struct PosteriorGP{Tprior,Tdata} <: AbstractGP
     data::Tdata
 end
 
+struct ExactGP end
+
+posterior(::ExactGP, fx::FiniteGP, y::AbstractVector{<:Real}) = posterior(fx, y)
```
Comment (theogf) [inline src/exact_gpr_posterior.jl:8]: I would have gone the other way around?

But maybe that's too complicated...

Comment (Crown421) [reply src/exact_gpr_posterior.jl:8]: I find this order of arguments quite intuitive.

File: src/sparse_approximations.jl [hunk 2]
```diff
@@ -228,14 +245,18 @@ true
...
-function elbo(vfe::VFE, fx::FiniteGP, y::AbstractVector{<:Real})
+function approx_log_evidence(vfe::VFE, fx::FiniteGP, y::AbstractVector{<:Real})
     @assert vfe.fz.f === fx.f
-    _dtc, A = _compute_intermediates(fx, y, vfe.fz)
-    return _dtc - (tr_Cf_invΣy(fx, fx.Σy) - sum(abs2, A)) / 2
+    dtc_objective, A = _compute_intermediates(fx, y, vfe.fz)
+    return dtc_objective - (tr_Cf_invΣy(fx, fx.Σy) - sum(abs2, A)) / 2
 end
 
+elbo(vfe::VFE, fx, y) = approx_log_evidence(vfe, fx, y)  # TODO deprecate?
```
Comment (st--) [inline src/sparse_approximations.jl:254]: Should we deprecate `elbo` or leave it as is (for `VFE`)?

Comment (simsurace) [reply src/sparse_approximations.jl:254]: I like the more specific name as a synonym wherever it is applicable.

File: src/exact_gpr_posterior.jl (hunk 1 above)
Comment (st--) [reply src/exact_gpr_posterior.jl:8]: Does it make any difference in practice? (I kinda like it as it is but that might just be status-quo bias too...)

Comment (Crown421) [reply src/exact_gpr_posterior.jl:8]: In practice I think the answer is no. 
Within the code, there might be something to be said for consistency. Every `posterior` is defined with the 3-argument form, but the exact one gets a special alias.

File: src/sparse_approximations.jl [hunk 3]
```diff
@@ -13,7 +13,14 @@ struct VFE{Tfz<:FiniteGP}
     fz::Tfz
 end
 
-const DTC = VFE
+"""
+    DTC(fz::FiniteGP)
+
+Similar to `VFE`, but uses a different objective for `approx_log_evidence`.
```
Comment (st--) [inline src/sparse_approximations.jl:19]: Could maybe do with a better docstring but then it needs to be sorted out more thoroughly anyways (see #309) and I can't  think of what it should be right now, so would leave that for some other PR/person/time...

File: test/sparse_approximations.jl [hunk 4]
```diff
@@ -1,133 +1,136 @@
...
+            # Ensure that the elbo/dtc objective is close to the logpdf when appropriate.
+            @test approx_log_evidence(ApproxType(f(x)), fx, y) isa Real
+            @test approx_log_evidence(ApproxType(f(x)), fx, y) ≈ logpdf(fx, y)
+
+            if ApproxType === VFE
+                @test elbo(ApproxType(f(x)), fx, y) ==
+                    approx_log_evidence(ApproxType(f(x)), fx, y)
+                @test elbo(ApproxType(f(x .+ randn(rng, N_cond))), fx, y) < logpdf(fx, y)
+            end
+        end
```
Comment (st--) [inline test/sparse_approximations.jl:101]: this is the main change

File: src/exact_gpr_posterior.jl (hunk 1 above)
Comment (theogf) [reply src/exact_gpr_posterior.jl:8]: To be fair, there is only one posterior. Shouldn't VFE and DTC dispatched on approx_posterior?

Comment (st--) [reply src/exact_gpr_posterior.jl]: @theogf what do you mean? there is no `approx_posterior` method..

**Comment by st--** [comment]
Hello! What do you think needs doing/addressing/... before merging this ?

File: src/sparse_approximations.jl [hunk 5]
```diff
@@ -161,48 +170,56 @@ function update_posterior(f_post_approx::ApproxPosteriorGP{<:VFE}, fz::FiniteGP)
...
         Σy=f_post_approx.data.Σy,
     )
-    return ApproxPosteriorGP(VFE(fz_new), f_post_approx.prior, cache)
+    return ApproxPosteriorGP(
+        _update_approx(f_post_approx.approx, fz_new), f_post_approx.prior, cache
+    )
 end
 
+_update_approx(vfe::VFE, fz_new::FiniteGP) = VFE(fz_new)
+_update_approx(dtc::DTC, fz_new::FiniteGP) = DTC(fz_new)
```
Comment (st--) [inline src/sparse_approximations.jl:179]: is this the right way to handle this? if anyone can think of a better approach please say:)

File: src/exact_gpr_posterior.jl (hunk 1 above)
Comment (theogf) [reply src/exact_gpr_posterior.jl]: haha somehow in my mind there was a `approx_posterior` method. So yeah then it's back to

- Should we have a unique 3-args posterior API (where the 2-args default to `ExactInference`) ?
- Should we just have 2-args methods and have the GP wrapped (like in ApproximateGPs) ?

Comment (st--) [reply src/exact_gpr_posterior.jl]: ApproximateGPs has the 3-args posterior though, no wrapping?

File: src/exact_gpr_posterior.jl [hunk 6]
```diff
@@ -3,10 +3,16 @@ struct PosteriorGP{Tprior,Tdata} <: AbstractGP
     data::Tdata
 end
 
+struct ExactGPR end
```
[the remark this answers is not in the document]
Comment (willtebbutt) [reply src/exact_gpr_posterior.jl:6]: I'd prefer `ExactInference`

[the remark this answers is not in the document]
Comment (st--) [reply src/exact_gpr_posterior.jl]: done

**Review by willtebbutt — APPROVED** [review]

**Pull request merged**